### PR TITLE
Modify functionality of "find" command

### LINKED ISSUE #96: Implement find command to find more parameters
Add feature for find command to find name, find email, find address, and find tag

### PR DESCRIPTION
**Pull request by delishad21**
Current implementation of find is limited as it can only search for names. This prevents filtering through other details (emails, tags that indicate tutorial class, etc.)

Lets improve the implementation of find by allowing it to be used to filter through other details

Fixes #96

### CODE REVIEW

File: src/main/java/seedu/address/model/person/PersonDetailContainsKeywordPredicate.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    private final Prefix prefix;
+    private final String keyword;
+
+    /**
+     * Constructor for the PersonDetailContainsKeywordPredicate.
+     * @param prefix The prefix to indicate the detail of the person to be searched.
+     * @param keyword The keyword to be searched.
+     */
+    public PersonDetailContainsKeywordPredicate(Prefix prefix, String keyword) {
+        this.prefix = prefix;
```
Comment (Pughal77) [inline src/main/java/seedu/address/model/person/PersonDetailContainsKeywordPredicate.java:28]: I see that now it accepts a new parameter called `prefix` here.

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 2]
```diff
@@ -38,6 +38,27 @@ public static boolean containsWordIgnoreCase(String sentence, String word) {
...
+    public static boolean containsSubstringIgnoreCase(String sentence, String subString) {
+        requireNonNull(sentence);
+        requireNonNull(subString);
+
+        String preppedString = subString.trim();
+        checkArgument(!preppedString.isEmpty(), "Substring parameter cannot be empty");
+
+        return sentence.toLowerCase().contains(preppedString.toLowerCase());
+    }
+
```
Comment (Pughal77) [inline src/main/java/seedu/address/commons/util/StringUtil.java:61]: Nice job creating a function to see if a substrings exists in a string! Our functionality of find command would be much better. I see that it is has good JUnit tests as well

**Review by Zer0Legion — COMMENTED** [review]
LGTM. Tested and works on my machine. Using the Gibbs' Reflective Cycle, we can reflect on the find command by finding students we want to find, eventually finding ourselves in the process. The code is correct and works on my machine, enabling me, roleplaying as a head tutor, to be able to find a fictional student without error.

**Review by Pughal77 — COMMENTED** [review]
Nice Job implementing find command. Only necessary files were changed

File: src/main/java/seedu/address/commons/util/StringUtil.java [hunk 3]
```diff
@@ -38,6 +38,27 @@ public static boolean containsWordIgnoreCase(String sentence, String word) {
...
+     *       </pre>
+     * @param sentence cannot be null
+     * @param subString cannot be null, cannot be empty
+     */
+    public static boolean containsSubstringIgnoreCase(String sentence, String subString) {
+        requireNonNull(sentence);
+        requireNonNull(subString);
+
+        String preppedString = subString.trim();
+        checkArgument(!preppedString.isEmpty(), "Substring parameter cannot be empty");
```
Comment (jayllo-c) [inline src/main/java/seedu/address/commons/util/StringUtil.java:57]: Great to see that you have ensured that the substring parameter should not be empty here!

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 4]
```diff
@@ -1,28 +1,42 @@
...
+    public static final Prefix[] ACCEPTED_PREFIXES = {PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL,
+                                                      PREFIX_ADDRESS, PREFIX_TAG};
+
     public static final String COMMAND_WORD = "find";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all persons whose names contain any of "
-            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
-            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
-            + "Example: " + COMMAND_WORD + " alice bob charlie";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all contacts with a specified aspect"
```
Comment (jayllo-c) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:29]: Nice to see you simplified the Message_Usage to be more clear for users to understand.

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 5]
```diff
@@ -19,15 +18,33 @@ public class FindCommandParser implements Parser<FindCommand> {
...
+     * and returns the found prefix.
+     * @throws ParseException if the user input does not conform to the expected format
+     */
+    private Prefix extractPrefixForFindCommand(ArgumentMultimap argMultimap) throws ParseException {
+
+        if (argMultimap.verifySinglePrefix()) {
+            for (Prefix prefix : FindCommand.ACCEPTED_PREFIXES) {
+                if (argMultimap.getValue(prefix).isPresent()
+                    && !argMultimap.getValue(prefix).get().isEmpty()) {
+                    return prefix;
```
Comment (jayllo-c) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java:42]: I now see how the code extracts a prefix

File: src/test/java/seedu/address/commons/util/StringUtilTest.java
Comment (jayllo-c) [inline src/test/java/seedu/address/commons/util/StringUtilTest.java:1]: Well done on ensuring the code is comprehensively tested.

**Review by jayllo-c — COMMENTED** [review]
Great job with the implementation of the find command. From what I see, the code structure looks correct and I cannot find any issues on the functionality of the code. The code coverage is also well done. Looks good to me!

**Pull request merged**